### DISCUSSION TITLE
SAML SOAP Backchannel Single-logout

### DISCUSSION THREAD

**Discussion by lagess** [keycloak/keycloak#9548] [Ideas]
Hello,

We have a need to support SAML Single-logout through backchannel with SOAP binding:
- the service provider (SP) sends a logout request to Keycloak (with SOAP binding)
- Keycloak sends a logout request (with SOAP binding) to all SP for which a client session exists

After some analysis of what is offered by Keycloak, I came to the conclusion that:
- **Logout:** Keycloak supports SAML backchannel logout toward the SP, BUT through POST binding
- **SAML SOAP binding:** Keycloak already supports SOAP binding login requests coming from the SP, BUT not logout requests

We would like to work on a PR to add the support of the SOAP binding for the SAML single logout functionality.
First, I'd need some additional information on the current implementation:

_Is there any reason why the backchannel logout was implemented like this in Keycloak? 
It does not seem to comply with the SAML specification, as the only backchannel binding is the SOAP binding [1]_

Thanks!

[1] https://docs.oasis-open.org/security/saml/v2.0/saml-profiles-2.0-os.pdf, section 4.4.3

**Comment by lagess** [comment]
We (https://github.com/cloudtrust) need to have proper backchannel logout through SOAP binding soon. So we propose to start working on this feature, and collect your feedback as soon as we have something ready for you to look at (same we did while developing the Artifact binding feature).
@stianst @hmlnarik Does this make sense to you?

Thanks!

**Reply by hmlnarik** [reply]
@lagess That sounds extremely well, thank you for the offer.
At this moment, our SAML experts are overwhelmed by other work but as this would be an wonderful and perhaps straightforward addition (in comparison with SAML ARTIFACT), we would be happy to have it kicked off and will be happy to help with reviews.

I am cc-ing @mhajas to watch this ticket and be primary point of contact on Keycloak side.

**Reply by lagess** [reply]
Thanks for the feedback! We will start working on that asap, and keep you updated.

**Comment by hmlnarik** [comment]
> Is there any reason why the backchannel logout was implemented like this in Keycloak?
> It does not seem to comply with the SAML specification, as the only backchannel binding is the SOAP binding [1]

AFAIK, only historical, there is no support for SOAP in Keycloak SAML adapters so I believe it had to use POST. We would be happy to have at least server part done correctly.

**Comment by lagess** [comment]
I'm back on the topic, currently working on the implementation of SOAP binding for SAML logout.
According to the SAML specification, the frontchannel implies asynchronous binding (POST, Redirect and Artifact), whereas backchannel implies synchronous binding (SOAP) [http://www.oasis-open.org/committees/download.php/56782/sstc-saml-profiles-errata-2.0-wd-07.pdf, 4.4.3.1].
This means that in the current Keycloak version, there is no correct SAML backchannel logout.

Now, question: shall I clean the current implementation of backchannel logout and align it with the SAML specification (SOAP only), or shall I only add the correct way and keep the current behavior as an option?

**Reply by mhajas** [reply]
@lagess Thank you for pointing this out. You are correct, this is something that is not compliant with the standard. However, this would not be simple to fix. We would need to also make the adapter compatible and I am not sure how much effort it would require. We can create a follow-up issue and solve it separately.

**Comment by lagess** [comment]
@mhajas 
I've opened the enhancement issue (https://github.com/keycloak/keycloak/issues/16293) + PRs for the support of SOAP backchannel logout. Could you please have a look when you have some time?
As stated in the issue, I've only added the SOAP backchannel whenever the SAML client is configured with backchannel + has a SOAP binding logout URL set. If these conditions are not met, the current behavior is kept.
Thanks!

**Reply by mhajas** [reply]
Thank you very much for the changes @lagess! I have quite a lot on my plate right now, but I will try to have at least some initial look next week.

**Reply by codespearhead** [reply]
@lagess your PR was merged on Feb 27, 2023. Should this discussion be closed?

**Reply by lagess** [reply]
Indeed we can close the discussion, thanks!

**Reply by slominskir** [reply]
Worth mentioning this appears to only apply to SAML clients, not a SAML IdP: https://github.com/keycloak/keycloak/discussions/23752

**Comment by mhajas** [comment]
Thank you for the reminder to close this @codespearhead!
